File: backend/Services/Parsers/SlayTheSpireParser.cs
using System.Text.RegularExpressions;
using backend.Models;

namespace backend.Services.Parsers
{
    public class SlayTheSpireParser : BasicScoreResultParser
    {
        private ILogger<SlayTheSpireParser> _logger;

        public SlayTheSpireParser(ILogger<SlayTheSpireParser> logger) : base(logger)
        {
            _logger = logger;
        }

        public override bool CountWinner => true;
        public override string GameName => "Slay the Spire Daily Challenge";
        public override bool GolfScoring => false;
        public override string? HelpText => "Runs submitted after 7 PM ET count for tomorrow.\nEntry examples: \"sts 700\", \"Sts 700\", or \"Slay the Spire 700\".";
        protected override Regex Parser => new Regex($"(slay the spire|sts) (?<{ScoreGroup}>\\d+)", RegexOptions.IgnoreCase);
        public override string Url => "https://www.megacrit.com/";
        protected override string? ExtraContent => null;

        protected override string GetCleanResult(string result, Match parserResults)
        {
            if (!parserResults.Groups.ContainsKey(ScoreGroup))
            {
                return "Slay the Spire";
            }

            var score = parserResults.Groups[ScoreGroup].Value ?? "";

            return "Slay the Spire" + "\n" + score;
        }

        protected override DailyResult SetScore(DailyResult dailyResult, Match parserResults)
        {
            // the Slay the Spire daily challenges resets at midnight UTC, so the date of the daily result
            // should match the current UTC date.
            dailyResult.Date = DateTime.UtcNow.Date;

            return base.SetScore(dailyResult, parserResults);
        }
    }
}

File: lambdas/src/parsers/slaythespire-parser.ts
import { DailyResult } from '../models/daily-result';
import { BasicScoreResultParser } from './base/basic-score-parser';

export class SlayTheSpireParser extends BasicScoreResultParser {
  readonly countWinner = true;
  readonly gameName = 'Slay the Spire Daily Challenge';
  readonly golfScoring = false;
  readonly helpText = 'Runs submitted after 7 PM ET count for tomorrow.\nEntry examples: "sts 700", "Sts 700", or "Slay the Spire 700".';
  protected readonly extraContent = null;
  readonly url = 'https://www.megacrit.com/';
  
  // Matches "Slay the Spire\n293" or "sts 700"
  // C# regex: `(slay the spire|sts) (?<score>\d+)` - requires exactly one space
  // TS version allows multiple whitespace/newlines for flexibility with multiline input
  protected readonly parser = new RegExp(
    `(slay the spire|sts)[\\s\\n]+(?<${BasicScoreResultParser.SCORE_GROUP}>\\d+)`,
    'i'
  );
  
  protected getCleanResult(result: string, match: RegExpMatchArray): string {
    const scoreStr = this.getScoreFromMatch(match);
    return 'Slay the Spire\n' + (scoreStr || '');
  }
  
  protected setScore(dailyResult: DailyResult, match: RegExpMatchArray): DailyResult {
    // The Slay the Spire daily challenges reset at midnight UTC, so the date of the daily result
    // should match the current UTC date.
    dailyResult.date = new Date().toISOString().split('T')[0];
    
    return super.setScore(dailyResult, match);
  }
  
  private getScoreFromMatch(match: RegExpMatchArray): string | null {
    if (match.groups && match.groups[BasicScoreResultParser.SCORE_GROUP]) {
      return match.groups[BasicScoreResultParser.SCORE_GROUP];
    }
    return null;
  }
}
